crypto/evp/p_lib.c

```diff
@@ -100,43 +107,75 @@ int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from)
         return 0;
     }
 
-    if (from->ameth && from->ameth->param_copy)
+    /* Only use the routine for provided keys if both keys are provided */
+    if (to->pkeys[0].keymgmt != NULL && from->pkeys[0].keymgmt != NULL) {
+        int rv = evp_keymgmt_util_copy(to, (EVP_PKEY *)from,
+                                       OSSL_KEYMGMT_SELECT_ALL_PARAMETERS);
+
+        /*
+         * Because the provided keydata was update beyond what legacy key
+         * there may be, we must sever the link to the latter.
+         */
+        if (rv == 1)
+            evp_pkey_free_legacy(to);
+        return rv;
+    }
+    if (from->ameth != NULL && from->ameth->param_copy != NULL)
         return from->ameth->param_copy(to, from);
  err:
     return 0;
 }
 
 int EVP_PKEY_missing_parameters(const EVP_PKEY *pkey)
 {
-    if (pkey != NULL && pkey->ameth && pkey->ameth->param_missing)
-        return pkey->ameth->param_missing(pkey);
+    if (pkey != NULL) {
+        if (pkey->pkeys[0].keymgmt != NULL)
+            return !evp_keymgmt_util_has((EVP_PKEY *)pkey,
+                                         OSSL_KEYMGMT_SELECT_ALL_PARAMETERS);
+        else if (pkey->ameth != NULL && pkey->ameth->param_missing != NULL)
+            return pkey->ameth->param_missing(pkey);
+    }
     return 0;
 }
 
 int EVP_PKEY_cmp_parameters(const EVP_PKEY *a, const EVP_PKEY *b)
 {
+    /* Only use the routine for provided keys if both keys are provided */
+    if (a->pkeys[0].keymgmt != NULL && b->pkeys[0].keymgmt != NULL)
+        /* We must unconst them for provider caching reasons. */
+        return evp_keymgmt_util_match((EVP_PKEY *)a, (EVP_PKEY *)b,
+                                      OSSL_KEYMGMT_SELECT_ALL_PARAMETERS);
+
+    /* All legacy keys */
     if (a->type != b->type)
         return -1;
-    if (a->ameth && a->ameth->param_cmp)
+    if (a->ameth != NULL && a->ameth->param_cmp != NULL)
         return a->ameth->param_cmp(a, b);
     return -2;
```

Comment:
So are we unable to compare parameters between a legacy key and a provided key?

Reply:
Damn...

Originally, I went with the provided routines as soon as any of the keys had `ameth == NULL`...  the trouble came with a fresh new key, just created with `EVP_PKEY_new()`, being passed in, and ending up in provided space, which simply did handle this well.

So yeahok, I need to look more closely at that.  Obviously, this PR will stay WIP a little longer than I expected....

Reply:
Okie, I reworked this a little bit.  It was not as hard as I feared it would be...